nixos/pcscd: allow use without polkit

rnhmjoj opened the pull request:
###### Description of changes

The polkit support in pcsclite is entirely optional but package enables
it unconditionally and this breaks connecting to the pcscd daemon on
systems without polkit.

The fix is making this configurable and automatically disabling
`polkitSupport` when the polkit service is disabled.


###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested programs can connect to the daemon (with and without `polkitSupport`)
- [ ] Tested compilation of all packages that depend on this change
- [x] Tested basic functionality of all binary files
- [x] 22.11 Release Notes - shouldn't be necessary
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

rnhmjoj committed:
nixos/pcscd: allow use without polkit

The polkit support in pcsclite is entirely optional but package enables
it unconditionally and this breaks connecting to the pcscd daemon on
systems without polkit.

The fix is making this configurable and automatically disabling
`polkitSupport` when the polkit service is disabled.

SuperSandro2000 commented on pkgs/tools/security/pcsclite/default.nix:
```diff
@@ -10,6 +10,7 @@
 , polkit
 , systemdMinimal
 , IOKit
+, polkitSupport ? true
```
```suggestion
, polkitSupport ? stdenv.isLinux
```

rnhmjoj replied:
Actually polkit and dbus should be portable: the package meta says `platforms.unix`.

amjoseph-nixpkgs replied:
> Actually polkit and dbus should be portable: the package meta says `platforms.unix`.

Do people really use dbus on Darwin?

(sorry if this is a stupid question; I got rid of my last Mac ~5 years ago)

rnhmjoj replied:
> Do people really use dbus on Darwin?

No idea, but if it (supposedly) works why not? Also, there's BSD too.

SuperSandro2000 replied:
Right now this enables polkit unconditional on darwin which it didn't before and which currently does not work. I would suggest to accept my change so that someone else can worry about polkit support on darwin and the build does not get broken.

rnhmjoj replied:
On second thought, it will probably have the same issue as on linux if the daemon not running. Maybe polkit should be disabled by default.

amjoseph-nixpkgs commented on pkgs/tools/security/pcsclite/default.nix:
```diff
@@ -34,14 +35,12 @@ stdenv.mkDerivation rec {
-    "--enable-polkit"
     "--with-systemdsystemunitdir=${placeholder "bin"}/lib/systemd/system"
-  ] else [
-    "--disable-libsystemd"
```
Isn't this deletion going to break Darwin builds?

rnhmjoj replied:
Notice the `lib.enableFeature stdenv.isLinux "systemd"` above: it should be equivalent but cleaner..

amjoseph-nixpkgs replied:
Ah, I see.

rnhmjoj commented:
@GrahamcOfBorg build pcsclite

rnhmjoj committed:
polkit: mark as linux-only

While polkit is supposedly cross-platform, the test suite requires
a Linux-specific package (to test the D-Bus functionality). Since the
build fails when disabling the test, this effectively limits the
platform to just Linux.

rnhmjoj commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9917,6 +9917,11 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) IOKit;
   };
 
+  pcscliteWithPolkit = pcsclite.override {
```
I'm not a fan of adding new top level names like this, but I don't know a simpler solution.
Ideally pcscd should detect/use polkit at runtime, but it doesn't, so we should provide both variants.

SuperSandro2000 commented on nixos/modules/services/hardware/pcscd.nix:
```diff
@@ -5,6 +5,10 @@ with lib;
   cfgFile = pkgs.writeText "reader.conf" config.services.pcscd.readerConfig;
 
+  package = if config.security.polkit.enable
+              then pkgs.pcscliteWithPolkit
```
```suggestion
              then (pkgs.pcsclite.override { polkitSupport = true; })
```

rnhmjoj replied:
Will this be built and distributed in binary cache?

rnhmjoj replied:
The [answer](https://matrix.to/#/%23dev%3Anixos.org/%24FZSSO-Q7LyfhpxF1kJR6eQBm5ZOog9_I2cmZRYM8PI4) is no.

SuperSandro2000 replied:
Can't we set the polkit support by default to on and disable it when polkit is disabled? Then most people get a cache hit and we don't need to add another top-level entry just for caching.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9917,6 +9917,11 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) IOKit;
   };
 
+  pcscliteWithPolkit = pcsclite.override {
+    pname = "pcsclite-with-polkit";
+    polkitSupport = true;
+  };
+
```
```suggestion
```

SuperSandro2000 reviewed (commented):
this would be a solution without a top-level entry

amjoseph-nixpkgs commented on pkgs/tools/security/pcsclite/default.nix:
```diff
@@ -10,6 +10,7 @@
 , polkit
 , systemdMinimal
 , IOKit
+, polkitSupport ? false
```
I'm no fan of polkit, but having this default to `false` is a rather major user-visible change, no?

Your PR, minus this latest commit, was looking really good and uncontroversial.  Why not leave it at that, and move the defaults-changing/package-adding into a separate subsequent PR?

rnhmjoj replied:
True, but the fact is polkit support makes pcscd unusable if you don't have polkit running.
The polkit support was added fairly recently (2b93e96d0bdf5cd98c6d7bd60af6e4dded4ab3d1) and it only caused trouble as far as I can tell: in addition to this issue it also formed a dependency cycle with dbus...

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9917,6 +9917,11 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) IOKit;
   };
 
+  pcscliteWithPolkit = pcsclite.override {
+    pname = "pcsclite-with-polkit";
+    polkitSupport = true;
+  };
```
```suggestion
  pcscliteWithPolkit = pcsclite.override { polkitSupport = true; };
```

rnhmjoj replied:
I think it's better to do it from all-packages.nix so one can easily see the name has changed.

SuperSandro2000 commented on pkgs/tools/security/pcsclite/default.nix:
```diff
@@ -10,6 +10,7 @@
+, polkitSupport ? false
 }:
 
 stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation rec {
  pname = "pcsclite" + lib.optionalString "-with-polkit";
```
or we leave this out entirely which I would prefer since people tend to use pname wherever they find a string matching it.

amjoseph-nixpkgs replied:
I think you mean

```suggestion
  pname = "pcsclite" + lib.optionalString polkitSupport "-with-polkit";
```

rnhmjoj replied:
> or we leave this out entirely which I would prefer since people tend to use pname wherever they find a string matching it.

This is necessary for nix-env -q/nix search not to list (apparently) duplicated packages: https://discourse.nixos.org/t/types-of-package-duplication-in-nix-search/6631/3

SuperSandro2000 commented on nixos/modules/services/hardware/pcscd.nix:
```diff
@@ -5,6 +5,10 @@ with lib;
 let
   cfgFile = pkgs.writeText "reader.conf" config.services.pcscd.readerConfig;
 
+  package = if config.security.polkit.enable
+              then pkgs.pcscliteWithPolkit
+              else pkgs.pcsclite;
```
```suggestion
  package = if config.security.polkit.enable then
    pkgs.pcscliteWithPolkit
  else
    pkgs.pcsclite;
```

SuperSandro2000 reviewed (commented):
oh well, not nice but hopefully more functional. With that we can merge it from my side.

SuperSandro2000 commented:
@ofborg eval
